pkg/storegateway/indexheader/threadpool.go

```diff
@@ -0,0 +1,71 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/cpg1111/threadpool-go/blob/master/threadpool.go
+// Provenance-includes-license: MIT
+// Provenance-includes-copyright: Christian Grabowski
+
+package indexheader
+
+import (
+	"context"
+	"errors"
+	"runtime"
+)
+
+type Threadpool struct {
+	ctx        context.Context
+	cancel     context.CancelFunc
+	pool       chan *OSThread
+	numThreads int
+}
+
+func NewThreadPool(ctx context.Context, cancel context.CancelFunc, num int) (*Threadpool, error) {
+	if num <= 0 {
+		return nil, nil
+	}
+
+	if num >= runtime.GOMAXPROCS(0) {
```

Comment:
To clarify - this is to check if the threadpool is going to be >= number of CPUs that can be executing simultaneously, yes?

Reply:
Correct. If the size of the threadpool is too large the service won't be able to run anything since the pool makes exclusive use of N threads.

Reply:
I think we should remove this limitation:

> If the size of the threadpool is too large the service won't be able to run anything since the pool makes exclusive use of N threads.

GOMAXPROCS != maximum number of threads our app can run. GOMAXPROCS is the number of threads where golang schedulers run (TIL this).

TL;DR, you can run 2*GOMAXPROCS thread-locked goroutines, and then run another one: https://go.dev/play/p/WFV3tysb4S5

Reply:
Nice! TIL

Reply:
(Me too :joy:)